### PR TITLE
datapath: Create MAC_BY_IFINDEX_MACRO in Go

### LINKED ISSUE #15246: CI: K8sServicesTest Checks service across nodes Tests NodePort BPF Tests with Wireguard (L2-less) Tests NodePort BPF
https://jenkins.cilium.io/job/Cilium-PR-K8s-1.13-net-next/955/testReport/Suite-k8s-1/13/K8sServicesTest_Checks_service_across_nodes_Tests_NodePort_BPF_Tests_with_Wireguard__L2_less__Tests_NodePort_BPF/

[test_results_Cilium-PR-K8s-1.13-net-next_955_BDD-Test-PR.zip](https://github.com/cilium/cilium/files/6102482/test_results_Cilium-PR-K8s-1.13-net-next_955_BDD-Test-PR.zip)

```
Stacktrace
/home/jenkins/workspace/Cilium-PR-K8s-1.13-net-next/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:518
Timed out after 240.000s.
Timeout while waiting for Cilium to become ready
Expected
    <*errors.errorString | 0xc0002a6d50>: {
        s: "only 0 of 2 desired pods are ready",
    }
to be nil
/home/jenkins/workspace/Cilium-PR-K8s-1.13-net-next/src/github.com/cilium/cilium/test/k8sT/assertionHelpers.go:126
Standard Output
Number of "context deadline exceeded" in logs: 0
Number of "level=error" in logs: 0
Number of "level=warning" in logs: 0
Number of "Cilium API handler panicked" in logs: 0
Number of "Goroutine took lock for more than" in logs: 0
No errors/warnings found in logs
Number of "context deadline exceeded" in logs: 0
Number of "level=error" in logs: 0
Number of "level=warning" in logs: 0
Number of "Cilium API handler panicked" in logs: 0
Number of "Goroutine took lock for more than" in logs: 0
No errors/warnings found in logs
Number of "context deadline exceeded" in logs: 0
⚠️  Number of "level=error" in logs: 8
⚠️  Number of "level=warning" in logs: 276
Number of "Cilium API handler panicked" in logs: 0
Number of "Goroutine took lock for more than" in logs: 0
Top 5 errors/warnings:
Command execution failed
Error while initializing daemon
++ grep ether
++ awk '{print $2}'
Unable to update ipcache map entry on pod add
Cilium pods: [cilium-4r2gz cilium-psmwt]
Netpols loaded: 
CiliumNetworkPolicies loaded: 
Endpoint Policy Enforcement:
Pod                           Ingress   Egress
testclient-t9clt                        
testclient-zmp5h                        
testds-4hqdh                            
testds-g9rrr                            
coredns-7ff984754c-gg997                
grafana-b4dbb994f-f46gd                 
prometheus-688959f59d-kq5z5             
test-k8s2-749b7c58cc-ddknw              

Standard Error
07:17:04 STEP: Running BeforeAll block for EntireTestsuite K8sServicesTest Checks service across nodes Tests NodePort BPF Tests with Wireguard (L2-less)
07:17:04 STEP: WaitforPods(namespace="kube-system", filter="-l app=kube-wireguarder")
07:17:19 STEP: WaitforPods(namespace="kube-system", filter="-l app=kube-wireguarder") => <nil>
07:17:19 STEP: SNAT with direct routing device wg0
07:17:19 STEP: Installing Cilium
07:17:21 STEP: Waiting for Cilium to become ready
FAIL: Timed out after 240.000s.
Timeout while waiting for Cilium to become ready
Expected
    <*errors.errorString | 0xc0002a6d50>: {
        s: "only 0 of 2 desired pods are ready",
    }
to be nil
=== Test Finished at 2021-03-08T07:21:21Z====
07:21:21 STEP: Running JustAfterEach block for EntireTestsuite K8sServicesTest
===================== TEST FAILED =====================
07:21:21 STEP: Running AfterFailed block for EntireTestsuite K8sServicesTest
cmd: kubectl get pods -o wide --all-namespaces
Exitcode: 0 
Stdout:
 	 NAMESPACE           NAME                               READY   STATUS             RESTARTS   AGE     IP              NODE   NOMINATED NODE   READINESS GATES
	 cilium-monitoring   grafana-b4dbb994f-f46gd            1/1     Running            0          34m     10.0.0.177      k8s1   <none>           <none>
	 cilium-monitoring   prometheus-688959f59d-kq5z5        1/1     Running            0          34m     10.0.0.75       k8s1   <none>           <none>
	 default             test-k8s2-749b7c58cc-ddknw         2/2     Running            0          22m     10.0.1.2        k8s2   <none>           <none>
	 default             testclient-t9clt                   1/1     Running            0          22m     10.0.1.232      k8s2   <none>           <none>
	 default             testclient-zmp5h                   1/1     Running            0          22m     10.0.0.86       k8s1   <none>           <none>
	 default             testds-4hqdh                       2/2     Running            0          21m     10.0.1.49       k8s2   <none>           <none>
	 default             testds-g9rrr                       2/2     Running            0          12m     10.0.0.82       k8s1   <none>           <none>
	 kube-system         cilium-4r2gz                       0/1     CrashLoopBackOff   5          4m3s    192.168.36.12   k8s2   <none>           <none>
	 kube-system         cilium-operator-65c988777b-tgw8n   1/1     Running            0          4m3s    192.168.36.13   k8s3   <none>           <none>
	 kube-system         cilium-psmwt                       0/1     CrashLoopBackOff   5          4m3s    192.168.36.11   k8s1   <none>           <none>
	 kube-system         coredns-7ff984754c-gg997           1/1     Running            0          33m     10.0.1.99       k8s2   <none>           <none>
	 kube-system         etcd-k8s1                          1/1     Running            0          37m     192.168.36.11   k8s1   <none>           <none>
	 kube-system         kube-apiserver-k8s1                1/1     Running            0          37m     192.168.36.11   k8s1   <none>           <none>
	 kube-system         kube-controller-manager-k8s1       1/1     Running            0          37m     192.168.36.11   k8s1   <none>           <none>
	 kube-system         kube-scheduler-k8s1                1/1     Running            0          37m     192.168.36.11   k8s1   <none>           <none>
	 kube-system         kube-wireguarder-cxl7r             1/1     Running            0          4m19s   192.168.36.12   k8s2   <none>           <none>
	 kube-system         kube-wireguarder-n7dtk             1/1     Running            0          4m19s   192.168.36.13   k8s3   <none>           <none>
	 kube-system         kube-wireguarder-t4ljf             1/1     Running            0          4m19s   192.168.36.11   k8s1   <none>           <none>
	 kube-system         log-gatherer-c76d8                 1/1     Running            0          34m     192.168.36.12   k8s2   <none>           <none>
	 kube-system         log-gatherer-ptc9g                 1/1     Running            0          34m     192.168.36.13   k8s3   <none>           <none>
	 kube-system         log-gatherer-x5lkj                 1/1     Running            0          34m     192.168.36.11   k8s1   <none>           <none>
	 kube-system         registry-adder-9d98r               1/1     Running            0          35m     192.168.36.11   k8s1   <none>           <none>
	 kube-system         registry-adder-rsn97               1/1     Running            0          35m     192.168.36.13   k8s3   <none>           <none>
	 kube-system         registry-adder-tmspd               1/1     Running            0          35m     192.168.36.12   k8s2   <none>           <none>
	 
Stderr:
 	 

Fetching command output from pods [cilium-4r2gz cilium-psmwt]
cmd: kubectl exec -n kube-system cilium-4r2gz -- cilium service list
Exitcode: 1 
Err: exit status 1
Stdout:
 	 
Stderr:
 	 error: unable to upgrade connection: container not found ("cilium-agent")
	 

cmd: kubectl exec -n kube-system cilium-4r2gz -- cilium endpoint list
Exitcode: 1 
Err: exit status 1
Stdout:
 	 
Stderr:
 	 error: unable to upgrade connection: container not found ("cilium-agent")
	 

cmd: kubectl exec -n kube-system cilium-psmwt -- cilium service list
Exitcode: 1 
Err: exit status 1
Stdout:
 	 
Stderr:
 	 error: unable to upgrade connection: container not found ("cilium-agent")
	 

cmd: kubectl exec -n kube-system cilium-psmwt -- cilium endpoint list
Exitcode: 1 
Err: exit status 1
Stdout:
 	 
Stderr:
 	 error: unable to upgrade connection: container not found ("cilium-agent")
	 

===================== Exiting AfterFailed =====================
07:21:26 STEP: Running AfterEach for block EntireTestsuite K8sServicesTest
07:21:26 STEP: Running AfterEach for block EntireTestsuite

[[ATTACHMENT|4cb677f1_K8sServicesTest_Checks_service_across_nodes_Tests_NodePort_BPF_Tests_with_Wireguard_(L2-less)_Tests_NodePort_BPF.zip]]
07:21:26 STEP: Running AfterAll block for EntireTestsuite K8sServicesTest Checks service across nodes Tests NodePort BPF Tests with Wireguard (L2-less)
 ```

### PR DESCRIPTION
**Pull request by brb**
See commit msgs.

Fix #15246.

### CODE REVIEW

**Comment by brb** [comment]
test-me-please

File: pkg/datapath/linux/config/config.go [hunk 1]
```diff
@@ -517,57 +513,58 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
...
-
-	if !anyL3Dev {
-		return "false", nil
-	}
-
-	return macro, nil
-}
-
-func macByIfIndexMacro() (string, error) {
+// devMacros generated NATIVE_DEV_MAC_BY_IFINDEX and IS_L3_DEV macros which
```
Comment (christarazi) [inline pkg/datapath/linux/config/config.go]: `generates`

**Review by joestringer — APPROVED** [review]
TIL `text/template`. LGTM :+1: 

I noted a similarity between the `CArrayString()` function introduced here and the existing `defineMAC()` (which uses two 32-bit values for easier templating/substitution), but given that these will be used to populate `union macaddr` structures I think it does make sense for both to exist in the tree.

**Comment by brb** [comment]
Previously, only the 4.9 test failed due to a flake.Pushed only a trivial comment change.

**Comment by brb** [comment]
test-4.9

**Comment by brb** [comment]
> I noted a similarity between the CArrayString() function introduced here and the existing defineMAC() (which uses two 32-bit values for easier templating/substitution), but given that these will be used to populate union macaddr structures I think it does make sense for both to exist in the tree.

Yep, unfortunately, we cannot reuse `CArrayString()` for the macaddr.

**Pull request merged**